db/external_sst_file_basic_test.cc

```diff
@@ -1901,27 +1901,181 @@ TEST_F(ExternalSSTFileBasicTest, AtomicReplaceData) {
     files.push_back(std::move(file4));
   }
 
+  const Snapshot* snapshot = db_->GetSnapshot();
+
   auto seqno_before_ingestion = db_->GetLatestSequenceNumber();
   ASSERT_OK(db_->IngestExternalFile(files, ifo));
   // Overlapping files each occupy one new sequence number.
   ASSERT_EQ(db_->GetLatestSequenceNumber(), seqno_before_ingestion + 3);
-  ASSERT_EQ(Get("a"), "a2");
-  ASSERT_EQ(Get("b"), "b2");
-  ASSERT_EQ(Get("x"), "x2");
-  ASSERT_EQ(Get("y"), "y2");
 
   // Check old version of data, big range deletion, new version of data are
   // on separate levels.
   ASSERT_EQ(2, NumTableFilesAtLevel(4));
   ASSERT_EQ(1, NumTableFilesAtLevel(5));
   ASSERT_EQ(2, NumTableFilesAtLevel(6));
 
-  CompactRangeOptions cro;
-  cro.bottommost_level_compaction = BottommostLevelCompaction::kForce;
-  ASSERT_OK(db_->CompactRange(cro, nullptr, nullptr));
+  ASSERT_OK(dbfull()->TEST_WaitForCompact());
+  ASSERT_EQ(2, NumTableFilesAtLevel(4));
+  ASSERT_EQ(1, NumTableFilesAtLevel(5));
+  ASSERT_EQ(2, NumTableFilesAtLevel(6));
+
+  bool compaction_iter_input_checked = false;
+  ROCKSDB_NAMESPACE::SyncPoint::GetInstance()->SetCallBack(
+      "VersionSet::MakeInputIterator:NewCompactionMergingIterator",
+      [&](void* arg) {
+        size_t* num_input_files = static_cast<size_t*>(arg);
+        EXPECT_EQ(1, *num_input_files);
+        compaction_iter_input_checked = true;
+      });
+  ROCKSDB_NAMESPACE::SyncPoint::GetInstance()->EnableProcessing();
+  db_->ReleaseSnapshot(snapshot);
+
+  ASSERT_OK(dbfull()->TEST_WaitForCompact());
+  ASSERT_EQ(2, NumTableFilesAtLevel(4));
+  ASSERT_EQ(0, NumTableFilesAtLevel(5));
+  ASSERT_EQ(0, NumTableFilesAtLevel(6));
+  ASSERT_TRUE(compaction_iter_input_checked);
+
+  ASSERT_EQ(Get("a"), "a2");
```

Comment:
I think we need overlapping key ranges but not completely overlapping key sets, between "old" and "new" data, in order to adequately test that the the range deletion is doing its job.

Reply:
Thank you Peter for the suggestion, do you mind give me an example?

Reply:
This test case `PartiallyReplaceDataWithOneStandaloneRangeDeletion` is updated to use not completely overlapping key sets between old and new data.